File: LeetCode/Hao/59SpiralMatrixII.java
public class Solution {
    private final static int[][] direction = new int[][] {
        {0, 1}, // turn right
        {1, 0}, // turn down
        {0, -1}, // turn left
        {-1, 0} // turn up
    };
    
    private boolean isInBound(int n, int x, int y) {
        return x >= 0 && x < n && y >= 0 && y < n;
    }
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int [n][n];
        int d = 0;
        int x = 0;
        int y = 0;
        int length = n * n;
        
        if (n == 0) {
            return matrix;
        }
        
        
        for (int i = 1; i <= length; i ++) {
            matrix[x][y] = i;
            
            for (int j = 0; j < 4; j ++) {
                int tmpx = x + direction[(d + j) % 4][0];
                int tmpy = y + direction[(d + j) % 4][1];
                
                if (isInBound(n, tmpx, tmpy) == true && matrix[tmpx][tmpy] == 0) {
                    d = (d + j) % 4;
                    x = tmpx;
                    y = tmpy;
                    
                    break;
                }
            }
        }
        
        return matrix;
    }
}

File: LeetCode/Lei/59SpiralMatrixII.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
      vector<vector<int>> matrix(n);
      if(n<=0)
        return matrix;
      for(int i=0; i<n; i++)
      {
        vector<int> v(n);
        matrix[i]=v;
      }
      int row=n, col=n;
      int r,c;
      int count=1;
      for(r=0,c=0; r<int((row+1)/2)&&c<int((col+1)/2); r++,c++)
      {
        //top
        for(int i=c; i<=(col-1)-c; i++)
        {
          matrix[r][i]=count;
          count++;
        }
        //right
        for(int i=r+1; i<=(row-1)-r; i++)
        {
          matrix[i][(col-1)-c]=count;
          count++;
        }
        //bottom
        for(int i=(col-1)-1-c; (row-1)-r>r&&i>=c; i--)
        {
          matrix[(row-1)-r][i]=count;
          count++;
        }
        //left
        for(int i=(row-1)-1-r; i>r&&(col-1)-1-c>=c; i--)
        {
          matrix[i][c]=count;
          count++;
        }
      }
        
    return matrix;
    }
};
